rhino/src/main/java/org/mozilla/javascript/AbstractEcmaObjectOperations.java

```diff
@@ -232,4 +242,65 @@ static void put(Context cx, Scriptable o, int p, Object v, boolean isThrow) {
             base.put(p, o, v);
         }
     }
+
+    /**
+     * Implement the ECMAScript abstract operation "GroupBy" defined in section 7.3.35 of ECMA262.
+     *
+     * @param cx
+     * @param scope
+     * @param items
+     * @param callback
+     * @param keyCoercion
+     * @see <a href="https://tc39.es/ecma262/#sec-groupby"></a>
+     */
+    static Map<Object, List<Object>> groupBy(
+            Context cx,
+            Scriptable scope,
+            IdFunctionObject f,
+            Object items,
+            Object callback,
+            KEY_COERCION keyCoercion) {
+        if (cx.getLanguageVersion() >= Context.VERSION_ES6) {
+            ScriptRuntimeES6.requireObjectCoercible(cx, items, f);
+        }
+        if (!(callback instanceof Callable)) {
+            throw ScriptRuntime.typeErrorById(
+                    "msg.isnt.function", callback, ScriptRuntime.typeof(callback));
+        }
+
+        // LinkedHashMap used to preserve key creation order
+        Map<Object, List<Object>> groups = new LinkedHashMap<>();
+        final Object iterator = ScriptRuntime.callIterator(items, cx, scope);
+        try (IteratorLikeIterable it = new IteratorLikeIterable(cx, scope, iterator)) {
+            double k = 0;
```

Comment:
Note to self for the future -- I wonder if there's a way that some day we can take this code and others like it and optimize it to use integers except in the rare case where we have more than 2^31 elements in an array. (And BTW I'm not so sure that we can handle more than 2^31 elements in an array anyway...)